hw-libs/drivers/SAMx70/d_time.c

```diff
@@ -33,125 +38,52 @@ void RTT_Handler(void)
 #if USE_FREERTOS
 		rtosResetTaskCounters();
 #endif
-		g_rollover++;
+		g_ticks.rollcnt++;
 	}
 }
 
 void time_init(void)
 {
-	static int initialized = 0;	if (initialized) { return; } initialized = 1;
+	// Only allow init once every reset
+	static bool inited = false;	
+	if (inited) return;
+	inited = true;
 
 #if USE_TIMER_DRIVER
 	timer_time_init();
 #endif
 
-	// configure RTT to increment as frequently as possible
+	// Increment as frequently as possible
 	rtt_sel_source(RTT, false);
 
-	rtt_init(RTT, RTPRES);
+	rtt_init(RTT, TIM_PRESC);
 	
-	// notify us when we rollover
-#if 0	// Bit shift 15 for faster rollover, works once.
-	rtt_write_alarm_time(RTT, 0x1FFFF);
-#else	// Standard operation
+	// Interrupt on 32bit rollover
 	rtt_write_alarm_time(RTT, 0);
-#endif
-	
 	rtt_enable_interrupt(RTT, RTT_MR_ALMIEN);
 	
 	NVIC_DisableIRQ(RTT_IRQn);
 	NVIC_ClearPendingIRQ(RTT_IRQn);
 	NVIC_SetPriority(RTT_IRQn, INT_PRIORITY_RTT);
 	NVIC_EnableIRQ(RTT_IRQn);
-	
-	g_rollover = 0;	// Reset the rollover now that the timer is fully configured.
 }
 
-inline volatile uint64_t time_ticks(void)
+inline volatile uint64_t time_ticks_u64(void)
 {
-	// Time must be read TWICE in ASF code and compared for corruptness.
-	
-	volatile uint32_t timer = RTT->RTT_VR;
-//	volatile uint64_t timer = RTT->RTT_VR;
-
-	while (timer != RTT->RTT_VR) 
-	{
-		timer = RTT->RTT_VR;
-	}
-
-#if 0	// Bit shift the timer to make overflow happen quicker (testing)
-	timer = timer << 15;
-#endif
+	// Time must be read until two consecutive reads match so we don't get a partially updated value
+	do {
+		g_ticks.tim = RTT->RTT_VR;
+	} while (g_ticks.tim != RTT->RTT_VR);
```

Comment:
This is a much more efficient implementation versus all the unnecessary copy logic I got rid of.